### PR TITLE
Elizabeth/trie cr

### PR DESCRIPTION
**Pull request by noot**
## Changes
- implemented basic trie. no caching, hashing, or db storage; just the data structure is done
- slightly modified keyToHex to take into account branch indexing
- preliminary memory db, doesn't do anything. will replace with badger DB in next pr

## Tests:
```
go test ./trie/...
```

### CODE REVIEW

**Comment by noot** [comment]
sorry about force push, there was a bunch of old commits that i wanted to get rid of. also not sure if the codeclimate issue is valid, but let me know what you think

**Comment by noot** [comment]
@ansermino insert has been refactored into separate insertBranch and insertExtension functions

File: trie/db.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package trie
+
+//import "sync"
+
+// Database is a high-level wrapper around a db
+// TODO: turn this into an actual database interface
+type Database struct {
+	memDB *MemDB
+}
```
Comment (dutterbutter) [inline trie/db.go]: We can remove this struct and reference #80

File: trie/node.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package trie
+
+type node interface {
+	// TODO, will be filled when we implement caching and hashing
+}
```
Comment (dutterbutter) [inline trie/node.go]: If its not being used we can just remove it for now

File: trie/trie.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	default:
+		err = errors.New("get error: invalid node")
+	}
+	return value, err
+}
+
+// Delete removes any existing value for key from the trie.
+func (t *Trie) Delete(key []byte) {
+	// TODO
+}
```
Comment (dutterbutter) [inline trie/trie.go]: I would remove until completed

Comment (noot) [reply trie/trie.go]: I think we should leave it, but not really sure if we've made any decisions regarding stub functions

File: trie/node.go (hunk 2 above)
Comment (noot) [reply trie/node.go]: this is required for compilation

File: trie/db.go (hunk 1 above)
Comment (noot) [reply trie/db.go]: okay cool will change that

**Pull request merged**